packages/keyring-eth-hd/src/hd-keyring.ts

```diff
@@ -0,0 +1,590 @@
+   * The message must be encrypted using the public key of the account.
+   * This method is compatible with the `eth_decryptMessage` RPC method
+   *
+   * @param withAccount - The address of the account.
+   * @param encryptedData - The encrypted data.
+   * @returns The decrypted message.
+   */
+  async decryptMessage(
+    withAccount: string,
+    encryptedData: EthEncryptedData,
+  ): Promise<string> {
+    const wallet = this.#getWalletForAccount(withAccount);
+    const { privateKey: privateKeyAsUint8Array } = wallet;
+    assert(privateKeyAsUint8Array, 'Expected private key to be set');
+    const privateKeyAsHex = Buffer.from(privateKeyAsUint8Array).toString('hex');
+    return decrypt({ privateKey: privateKeyAsHex, encryptedData });
+  }
+
+  /**
+   * Sign a typed message using the private key of the specified account.
+   * This method is compatible with the `eth_signTypedData` RPC method.
+   *
+   * @param withAccount - The address of the account.
+   * @param typedData - The typed data to sign.
+   * @param opts - The options for signing the message.
+   * @returns The signature of the message.
+   */
+  async signTypedData<Types extends MessageTypes>(
+    withAccount: string,
+    typedData: TypedDataV1 | TypedMessage<Types>,
+    opts: HDKeyringAccountSelectionOptions & {
+      version: SignTypedDataVersion;
+    } = { version: SignTypedDataVersion.V1 },
+  ): Promise<string> {
+    // Treat invalid versions as "V1"
+    const version = Object.keys(SignTypedDataVersion).includes(opts.version)
+      ? opts.version
+      : SignTypedDataVersion.V1;
+
+    const privateKey = this.#getPrivateKeyFor(withAccount, opts);
+    return signTypedData({
+      privateKey: Buffer.from(privateKey),
+      data: typedData,
+      version,
+    });
+  }
+
+  /**
+   * Remove an account from the keyring.
+   *
+   * @param account - The address of the account to remove.
+   */
+  removeAccount(account: string): void {
+    const address = normalize(account);
+    assert(address, 'Must specify a valid address.');
+    if (
+      !this.#wallets
+        .map(
+          ({ publicKey }) => publicKey && this.#addressfromPublicKey(publicKey),
+        )
+        .includes(address)
+    ) {
+      throw new Error(`Address ${address} not found in this keyring`);
+    }
+
+    this.#wallets = this.#wallets.filter(
+      ({ publicKey }) =>
+        publicKey && this.#addressfromPublicKey(publicKey) !== address,
+    );
+  }
+
+  /**
+   * Get the public key of the account to be used for encryption.
+   *
+   * @param withAccount - The address of the account.
+   * @param opts - The options for selecting the account.
+   * @returns The public key of the account.
+   */
+  async getEncryptionPublicKey(
+    withAccount: string,
+    opts: HDKeyringAccountSelectionOptions = {},
+  ): Promise<string> {
+    const privKey = this.#getPrivateKeyFor(withAccount, opts);
+    const publicKey = getEncryptionPublicKey(bytesToHex(privKey));
```

Comment:
In the original implementation `privKey` was passed to `getEncryptionPublicKey` directly:
```typescript
const publicKey = getEncryptionPublicKey(privKey);
```
Though, the addition of types evidenced that `getEncryptionPublicKey` only accept hex strings, hence the conversion with `bytesToHex`